File: .config/nvim/lua/config/plugins/ddu.lua
local M = {
  "Shougo/ddu.vim",
  dependencies = {
    "vim-denops/denops.vim",

    --- ui ---
    "Shougo/ddu-ui-ff",
    "matsui54/ddu-vim-ui-select",

    --- source ---
    "4513ECHO/ddu-source-emoji",
    "4513ECHO/ddu-source-ghq",
    "4513ECHO/ddu-source-source",
    "Shougo/ddu-source-line",
    "Shougo/ddu-source-register",
    "matsui54/ddu-source-command_history",
    "shun/ddu-source-buffer",

    --- kind ---
    "4513ECHO/ddu-kind-url",
    "Shougo/ddu-kind-file",
    "Shougo/ddu-kind-word",

    ---- ddu-kind-url deps
    "tyru/open-browser.vim",

    --- matcher ---
    "Milly/ddu-filter-kensaku",

    ---- ddu-filter-kensaku deps
    "lambdalisue/kensaku.vim",
  },
}

function M.init()
  local start = vim.fn["ddu#start"]
  vim.keymap.set("n", "<leader>d", function() start({ sources = { { name = "source" } } }) end)
end

function M.config()
  local patch_global = vim.fn["ddu#custom#patch_global"]

  -- ui
  patch_global("ui", "ff")
  patch_global("uiParams", {
    ff = {
      split = "floating",
      floatingBorder = "single",
    },
  })

  -- sources
  patch_global("sources", {})
  patch_global("sourceOptions", {
    ["_"] = { matchers = { "matcher_kensaku" } },
  })
  patch_global("sourceParams", {})

  --- kinds
  patch_global("kindOptions", {
    command_history = { defaultAction = "edit" },
    file = { defaultAction = "open" },
    source = { defaultAction = "execute" },
    ui_select = { defaultAction = "select" },
    url = { defaultAction = "browse" },
    word = { defaultAction = "append" },
  })

  -- key-bindings
  vim.api.nvim_create_autocmd("FileType", {
    pattern = "ddu-ff",
    callback = function()
      local action = vim.fn["ddu#ui#ff#do_action"]
      local opts = { buffer = true }
      vim.keymap.set("n", "<CR>", function() action("itemAction") end, opts)
      vim.keymap.set("n", "<Space>", function() action("toggleSelectItem") end, opts)
      vim.keymap.set("n", "i", function() action("openFilterWindow") end, opts)
      vim.keymap.set("n", "q", function() action("quit") end, opts)
    end,
  })

  vim.api.nvim_create_autocmd("FileType", {
    pattern = "ddu-ff-filter",
    callback = function()
      local opts = { buffer = true }
      vim.keymap.set("i", "<CR>", "<Esc><Cmd>call ddu#ui#ff#close()<CR>", opts)
      vim.keymap.set("n", "<CR>", "<Cmd><Cmd>call ddu#ui#ff#close()<CR>", opts)
    end,
  })
end

return M

File: .config/nvim/dpprc/hooks/ddu.ts
import {
  BaseConfig,
  ContextBuilder,
} from "https://deno.land/x/ddu_vim@v3.6.0/types.ts";
import type { Denops } from "https://deno.land/x/ddu_vim@v3.6.0/deps.ts";
import type { Params as FfParams } from "https://deno.land/x/ddu_ui_ff@v1.1.0/ff.ts";
import type { Params as FilerParams } from "https://deno.land/x/ddu_ui_filer@v1.1.0/filer.ts";

const windowParams = {
  // window size and position
  winWidth: "&columns * 4 / 5",
  winHeight: "&lines * 4 / 5",
  winRow: "&lines / 10",
  winCol: "&columns / 10",
  previewWidth: "&columns * 2 / 5 - 1",
  previewHeight: "&lines * 4 / 5 - 2",
  previewRow: "&lines / 10 + 1",
  previewCol: "&columns / 2 - 1",

  // window decorations
  split: "floating",
  floatingBorder: "single",
  previewFloating: true,
  previewFloatingBorder: "single",
  previewSplit: "vertical",
  highlights: {
    floating: "Normal",
    floatingBorder: "Special",
  },
} as const satisfies Partial<FfParams> satisfies Partial<FilerParams>;

const dduSourceLSPParams = {
  clientName: "lspoints",
};

export class Config extends BaseConfig {
  // deno-lint-ignore require-await
  override async config(args: {
    denops: Denops;
    contextBuilder: ContextBuilder;
  }): Promise<void> {
    args.contextBuilder.patchLocal("lsp-definition", {
      sync: true,
      sources: [
        {
          name: "lsp_definition",
          params: { method: "textDocument/definition" },
        },
        {
          name: "lsp_definition",
          params: { method: "textDocument/typeDefinition" },
        },
        {
          name: "lsp_definition",
          params: { method: "textDocument/declaration" },
        },
        {
          name: "lsp_definition",
          params: { method: "textDocument/implementation" },
        },
      ],
      uiParams: {
        ff: {
          immediateAction: "open",
        } as Partial<FfParams>,
      },
    });
    args.contextBuilder.patchLocal("lsp-references", {
      sync: true,
      sources: [{ name: "lsp_references" }],
      uiParams: {
        ff: {
          immediateAction: "open",
        } as Partial<FfParams>,
      },
    });
    args.contextBuilder.patchLocal("lsp-diagnostic", {
      sync: true,
      sources: [{
        name: "lsp_diagnostic",
        params: {
          buffer: 0,
        },
      }],
    });
    args.contextBuilder.patchLocal("lsp-codeAction", {
      sync: true,
      sources: [{ name: "lsp_codeAction" }],
    });
    args.contextBuilder.patchLocal("git-ls-files", {
      sources: [{
        name: "file_external",
        params: { "cmd": ["git", "ls-files"] },
      }],
    });
    args.contextBuilder.patchLocal("ghq", {
      sources: [{ name: "ghq" }],
      kindOptions: {
        file: {
          defaultAction: "cd",
        },
      },
      uiParams: {
        ff: {
          startAutoAction: false,
        },
      },
    });
    args.contextBuilder.patchLocal("filer", {
      ui: "filer",
      uiParams: {
        filer: {
          ...windowParams,
        },
      },
      sources: [{ name: "file", params: {} }],
      sourceOptions: {
        _: {
          columns: ["filename"],
        },
      },
      kindOptions: {
        file: {
          defaultAction: "open",
        },
      },
    });
    args.contextBuilder.patchGlobal({
      ui: "ff",
      uiParams: {
        ff: {
          ...windowParams,
          filterSplitDirection: "floating",
          startAutoAction: false,
          autoAction: {
            name: "preview",
          },
        } as Partial<FfParams>,
      },

      sources: [],
      sourceParams: {
        "lsp_definition": {
          ...dduSourceLSPParams,
        },
        "lsp_references": {
          ...dduSourceLSPParams,
        },
        "lsp_documentSymbol": {
          ...dduSourceLSPParams,
        },
        "lsp_workspaceSymbol": {
          ...dduSourceLSPParams,
        },
        "lsp_callHierarchy": {
          ...dduSourceLSPParams,
        },
        "lsp_typeHierarchy": {
          ...dduSourceLSPParams,
        },
        "lsp_codeAction": {
          ...dduSourceLSPParams,
        },
        "lsp_diagnostic": {
          ...dduSourceLSPParams,
        },
      },
      sourceOptions: {
        _: {
          ignoreCase: true,
          matchers: ["matcher_substring"],
        },
        copilot: {
          defaultAction: "append",
        },
        emoji: {
          defaultAction: "append",
        },
        line: {
          matchers: ["matcher_kensaku"],
        },
      },

      kindOptions: {
        command_history: { defaultAction: "edit" },
        file: { defaultAction: "open" },
        source: { defaultAction: "execute" },
        ui_select: { defaultAction: "select" },
        url: { defaultAction: "browse" },
        word: { defaultAction: "append" },
        lsp: { defaultAction: "open" },
        lsp_codeAction: { defaultAction: "apply" },
        help: { defaultAction: "tabopen" },
        window: { defaultAction: "open" },
      },
    });
  }
}
